lib/Backends/CPU/CPUBackend.h

```diff
@@ -38,7 +38,9 @@ class CPUBackend : public LLVMBackend {
   ///@{
   virtual ~CPUBackend() override = default;
 
-  std::string getBackendName() const override { return getName(); }
+  std::string getBackendName() const override {
+    return Named::getName().empty() ? getName() : Named::getName().str();
```

Comment:
I'm not quite happy about having a static `getName` in each backend and a dynamic `Named::getName()` used to set the name dynamically. But I was hesitant to completely re-factor the existing registration machinery.

Reply:
Hmm. Is it possible to ditch the static `getName()` in the backends and instead always use `Named`? e.g. `Interpreter() : Named("Interpreter") = {};` and then we could always just use `Named::getName()`?

Reply:
Yes, I've thought about it and I'm not sure. I think it would interfere with the registration mechanism. Currently, we do this inside factory classes expanded from the macro:
```c++
std::string getRegistrationKey() const override { 
      return BackendClass::getName(); 
}
unsigned numDevices() const override { 
      return BackendClass::numDevices(); 
}   
```

It relies on the possibility of obtaining the name (and numDevices) of the backend without creating an instance of the backend. I don't know if we want to maintain this property. @rdzhabarov introduced this mechanism and probably knows reasons for that. 

If we decide it is fine to create an instance of the backend to obtain the name and number of devices, we can totally remove "static" from those methods.

Reply:
Ah right, OK. It seems like we could change the macro for `getRegistrationKey()` to return `#BackendClass`, though I suppose the class could then return a different name for `getName()`. I'm not sure we actually need `numDevices()` to be a static method, do we? Would this help?

Reply:
> It seems like we could change the macro for getRegistrationKey() to return #BackendClass,

The name of the BackendClass is e.g. CPUBackend, but the name of the backend is "CPU". 

The numDevices could probably be a non-static method, though DeviceManager::numDevices uses it through a backend factory as far as I can see.